(Draft) New Data Model (#480)
hermes-hmc/workflow: docs/source/dev/data_model.md

<!--
SPDX-FileCopyrightText: 2025 German Aerospace Center (DLR)

SPDX-License-Identifier: CC-BY-SA-4.0
-->

<!--
SPDX-FileContributor: Stephan Druskat <stephan.druskat@dlr.de>
-->

# Data model

`hermes`' internal data model acts like a contract between `hermes` and plugins.
It is based on [**JSON-LD (JSON Linked Data)**](https://json-ld.org/), and
the public API simplifies interaction with the data model through Python code.

> Output of the different `hermes` commands consequently is valid JSON-LD, serialized as JSON, that is cached in 

Reviewer: >valid JSON-LD, serialized as JSON

Obviously 😁 

```suggestion
Output of the different `hermes` subcommands consequently are valid JSON-LD files that are cached in 
```

> subdirectories of the `.hermes/` directory that is created in the root of the project directory.

Reviewer: Maybe add a little directory tree like this (but updated with current file names):

```
.hermes
├── harvest
│   ├── cff_contexts.json
│   ├── cff.json
│   ├── file_exists_contexts.json
│   └── file_exists.json
└── process
    ├── hermes.json
    └── tags.json
```

(`tree .hermes` in Linux or `tree /f .hermes` in Powershell on Windows)

> The cache is purely for internal purposes, its data should not be interacted with.

Reviewer: Maybe the other way around:

```suggestion
The cache should only be interacted with via the `hermes` libraries.
```

### JSON-LD for plugin developers

```{attention}
Work in progress.
```

### Working with the `hermes` data model in plugins 

> **Goal**  
> Understand how plugins access the `hermes` data model and interact with it.

`hermes` aims to hide as much of the data model as possible behind a public API
to avoid that plugin developers have to deal with some of the more complex features of JSON-LD.

#### Model instances in different types of plugin

You can extend `hermes` with plugins for three different commands: `harvest`, `curate`, `deposit`.

The commands differ in how they work with instances of the data model.

- `harvest` plugins _create_ a single new model instance and return it.
- `curate` plugins are passed a single existing model instance (the output of `process`),
and return a single model instance.
- `deposit` plugins are passed a single existing model instance (the output of `curate`),
and return a single model instance.

#### How plugins work with the API

```{important}
Plugins access the data model _exclusively_ through the API class {class}`hermes.model.SoftwareMetadata`.
```
 
The following sections show how this class works. 

> ##### Creating a data model instance

Reviewer: h5 is too many levels down. It's absolutely fine to have multiple paragraphs with multiple sentences each in a subdivision.

```{code-block} python
:caption: Initializing a default data model instance
from hermes.model import SoftwareMetadata

data = SoftwareMetadata()
```

`SoftwareMetadata` objects initialized without arguments provide the default _context_
(see [_JSON-LD for plugin developers_](#json-ld-for-plugin-developers)).
This means that now, you can use terms from the schemas included in the default context to describe software metadata.

Terms from [_CodeMeta_](https://codemeta.github.io/terms/) can be used without a prefix:

```{code-block} python
:caption: Using terms from the default schema
data["readme"] = ...
```

Terms from [_Schema.org_](https://schema.org/) can be used with the prefix `schema`:

```{code-block} python
:caption: Using terms from a non-default schema
data["schema:copyrightNotice"] = ...
```

You can also use other linked data vocabularies. To do this, you need to identify them with a prefix and register them
with the data model by passing it `extra_vocabs` as a `dict` mapping prefixes to URLs where the vocabularies are
provided as JSON-LD:

```{code-block} python
:caption: Injecting additional schemas
from hermes.model import SoftwareMetadata

> # Contents served at https://bar.net/schema.jsonld:

Reviewer: Please only use reserved example domains! https://en.wikipedia.org/wiki/Example.com

# {
#    "@context":
#    {
#       "name": "https://schema.org/name"
#    }
# }

data = SoftwareMetadata(extra_vocabs={"foo": "https://bar.net/schema.jsonld"})

data["foo:name"] = ...
```

##### Adding data

Once you have an instance of {class}`hermes.model.SoftwareMetadata`, you can add data to it,
i.e., metadata that describes software:

```{code-block} python
:caption: Setting data values
data["name"] = "My Research Software"  # A simple "Text"-type value
# → Simplified model representation : { "name": [ "My Research Software" ] }
# Cf. "Accessing data" below
data["author"] = {"name": "Shakespeare"}  # An object value that uses terms available in the defined context
# → Simplified model representation : { "name": [ "My Research Software" ], "author": [ { "name": "Shakespeare" } ] }
# Cf. "Accessing data" below
```

##### Accessing data

You need to be able to access data in the data model instance to add, edit or remove data.
Data can be accessed by using term strings, similar to how values in Python `dict`s are accessed by keys.

```{important}
When you access data from a data model instance,
it will always be returned in a **list**-like object!
```

> The reason for providing data in list-like objects is that JSON-LD treats all property values as arrays.

Reviewer: Link to the related section in the JSON-LD spec would be nice

        print(f"Author {i} has expected name.")
    else:
        raise ValueError("Unexpected author name found!", author["name"][0])

# Mock output:
# $> Author 1 has expected name.
# $> Author 2 has expected name.
```

```{code-block} python
:caption: Value check 
for email in data["author"][0]["email"]:
    if email.endswith(".edu"):
        print("Shakespeare has an email address at an educational institution.")
    else:
        print("Cannot confirm affiliation with educational institution for Shakespeare.")

# Mock output
# $> Cannot confirm affiliation with educational institution for author.
```

```{code-block} python
:caption: Value check and list comprehension 
if all(["hamilton" in email for email in data["author"][1]["email"]]):
    print("Author has only emails with their name in it.")

# Mock output
# $> Author has only emails with their name in it.
```

The example continues to show how to assert data values.

As mentioned in the [introduction to the data model](#data-model), 
`hermes` uses a JSON-LD-like internal data model. 
The API class {class}`hermes.model.SoftwareMetadata` hides many
of the more complex aspects of JSON-LD and makes it easy to work

> with the data model.

Reviewer: This feels like a misplaced summary. I would just remove it.

So the API class hides the internal model objects.
Therefore, they work as you would expect from plain
Python data:

```{code-block} python

> :caption: Naive containment assertion that raises

Reviewer: It doesn't though. The mock output is from the path where the assertion holds.

Reviewer 2: I'll look through the entire file and correct all other outdated information.

:emphasize-lines: 5,13 
try:
    assert (
            {'name': ['Shakespeare'], 'email': ['shakespeare@baz.net']}
            in
            data["author"]
    )
    print("The author was found!")
except AssertionError:
    print("The author could not be found.")
    raise

# Mock output
# $> The author was found!
#
#
# Internal Model from data["author"]:
#        {'@list': [
#            {
#                'http://schema.org/name': [{'@value': 'Shakespeare'}],
#                'http://schema.org/email': [{'@value': 'shakespeare@baz.net'}]
#            },
#            {
#                'http://schema.org/name': [{'@value': 'Hamilton'}],
#                'http://schema.org/email': [
#                    {'@list': [
#                        {'@value': 'hamilton@baz.net'}, {'@value': 'hamilton@spam.org'}, {'@value': 'hamilton@eggs.com'}
#                    ]}
#                ]
#            }]
#        }
#    )
```

---

## See Also

> - API reference: {class}`hermes.model.SoftwareMetadata`

Reviewer: This link never works throughout the whole docs. For some reason, the class is only exposed in the docs as `hermes.model.api.SoftwareMetadata`
